File: test_conversions/api_client.rb
require 'net/http'
require 'json'
require 'uri'

class ApiClient
  DEFAULT_TIMEOUT = 30

  def initialize(base_url, api_key: nil)
    @base_url = base_url.chomp('/')
    @api_key  = api_key
    @headers  = { 'Content-Type' => 'application/json' }
    @headers['Authorization'] = "Bearer #{@api_key}" if @api_key
  end

  def get(path, params: {})
    uri = build_uri(path, params)
    request = Net::HTTP::Get.new(uri)
    apply_headers(request)
    execute(uri, request)
  end

  def post(path, body: {})
    uri = build_uri(path)
    request = Net::HTTP::Post.new(uri)
    apply_headers(request)
    request.body = body.to_json
    execute(uri, request)
  end

  def delete(path)
    uri = build_uri(path)
    request = Net::HTTP::Delete.new(uri)
    apply_headers(request)
    execute(uri, request)
  end

  private

  def build_uri(path, params = {})
    uri = URI("#{@base_url}#{path}")
    uri.query = URI.encode_www_form(params) unless params.empty?
    uri
  end

  def apply_headers(request)
    @headers.each { |k, v| request[k] = v }
  end

  def execute(uri, request)
    Net::HTTP.start(uri.host, uri.port,
                    use_ssl: uri.scheme == 'https',
                    read_timeout: DEFAULT_TIMEOUT) do |http|
      response = http.request(request)
      {
        status: response.code.to_i,
        body:   JSON.parse(response.body, symbolize_names: true),
        ok:     response.code.to_i < 400
      }
    end
  rescue StandardError => e
    { status: 0, body: { error: e.message }, ok: false }
  end
end

client = ApiClient.new('https://api.example.com', api_key: 'secret')
result = client.get('/users', params: { page: 1, limit: 10 })
puts result[:status]
puts result[:ok]

File: test_conversions/api_client.py
import json
from urllib.parse import urljoin, urlencode
from http.client import HTTPSConnection, HTTPConnection
from dataclasses import dataclass

@dataclass
class Response:
    status: int
    body: dict
    ok: bool

class ApiClient:
    DEFAULT_TIMEOUT = 30

    def __init__(self, base_url, api_key=None):
        self.base_url = base_url.rstrip('/')
        self.api_key = api_key
        self.headers = {'Content-Type': 'application/json'}
        if api_key:
            self.headers['Authorization'] = f"Bearer {api_key}"

    def get(self, path, params=None):
        if params is None:
            params = {}
        uri = self.build_uri(path, params)
        return self.execute(uri, 'GET')

    def post(self, path, body=None):
        if body is None:
            body = {}
        uri = self.build_uri(path)
        return self.execute(uri, 'POST', body=json.dumps(body))

    def delete(self, path):
        uri = self.build_uri(path)
        return self.execute(uri, 'DELETE')

    def build_uri(self, path, params=None):
        if params is None:
            params = {}
        uri = urljoin(self.base_url, path)
        if params:
            uri += '?' + urlencode(params)
        return uri

    def execute(self, uri, method, body=None):
        try:
            if uri.startswith('https'):
                conn = HTTPSConnection(uri.split('://')[-1].split('/')[0], timeout=self.DEFAULT_TIMEOUT)
            else:
                conn = HTTPConnection(uri.split('://')[-1].split('/')[0], timeout=self.DEFAULT_TIMEOUT)
            conn.request(method, uri.split('://')[-1], headers=self.headers, body=body)
            response = conn.getresponse()
            body = json.loads(response.read())
            return Response(response.status, body, response.status < 400)
        except Exception as e:
            return Response(0, {'error': str(e)}, False)
        finally:
            if 'conn' in locals():
                conn.close()

client = ApiClient('https://api.example.com', api_key='secret')
result = client.get('/users', params={'page': 1, 'limit': 10})
print(result.status)
print(result.ok)
